blaze-core/src/main/scala/org/http4s/blaze/websocket/Http4sWSStage.scala

```diff
@@ -0,0 +1,111 @@
+
+class Http4sWSStage(ws: ws4s.Websocket)(implicit val strategy: Strategy) extends TailStage[WebSocketFrame] {
+  def name: String = "Http4s WebSocket Stage"
+
+  // TODO Remove
+  def log[A](prefix: String): Pipe[Task, A, A] = _.evalMap{a => Task.delay {println(s"$prefix> $a"); a} }
+
+  private val deadSignal: Task[Signal[Task, Boolean]] = async.signalOf[Task, Boolean](false)
+
+  //////////////////////// Source and Sink generators ////////////////////////
+
+  def snk: Sink[Task, WebSocketFrame] = _.evalMap { frame =>
+    Task.async[Unit] { cb =>
+      channelWrite(frame).onComplete {
+        case Success(res)             => cb(Right(res))
+        case Failure(t @ Command.EOF) => cb(Left(t))
+        case Failure(t)               => cb(Left(t))
+      }(directec)
+    }
+  }
+
+  def inputstream: Stream[Task, WebSocketFrame] = {
+    val t = Task.async[WebSocketFrame] { cb =>
+      def go(): Unit = channelRead().onComplete {
+        case Success(ws)         => ws match {
+            case Close(_)    =>
+              for {
+                t <- deadSignal.map(_.set(true))
+              } yield {
+                t.unsafeRun()
+                cb(Left(Command.EOF))
+              }
+
+            // TODO: do we expect ping frames here?
+            case Ping(d)     =>  channelWrite(Pong(d)).onComplete {
+              case Success(_)           => go()
+              case Failure(Command.EOF) => cb(Left(Command.EOF))
+              case Failure(t)           => cb(Left(t))
+            }(trampoline)
+
+            case Pong(_)     => go()
+            case f           => cb(Right(f))
+          }
+
+        case Failure(Command.EOF) => cb(Left(Command.EOF))
+        case Failure(e)           => cb(Left(e))
+      }(trampoline)
+
+      go()
+    }
+    Stream.repeatEval(t)
+  }
+
+  //////////////////////// Startup and Shutdown ////////////////////////
+
+  override protected def stageStartup(): Unit = {
+    super.stageStartup()
+
+    // A latch for shutting down if both streams are closed.
+    val count = new java.util.concurrent.atomic.AtomicInteger(2)
+
+    // If both streams are closed set the signal
+    val onStreamFinalize: Task[Unit] =
+      for {
+        dec <- Task.delay(count.decrementAndGet())
+        _   <- deadSignal.map(signal => if (dec == 0) signal.set(true))
+      } yield ()
+
+    // Task to send a close to the other endpoint
+    val sendClose: Task[Unit] = Task.delay(sendOutboundCommand(Command.Disconnect))
+
+    val wsStream = for {
+      dead   <- deadSignal
+      in     = inputstream.through(log("input")).to(ws.write).onFinalize(onStreamFinalize)
+      out    = ws.read.through(log("output")).onFinalize(onStreamFinalize).to(snk).drain
+      merged <- (in mergeHaltR out).interruptWhen(dead).onFinalize(sendClose).run
+    } yield merged
+
+    wsStream.or(sendClose).unsafeRunAsync {
+      case Left(_)  => sendClose.unsafeRun() // RFC How to avoid this call?
```

Comment:
Right, this is probably wrong. That left should be raised if the previous `sendClose` failed and in that case there isn't much to do. I think the only thing we can do is to log the error. WDYT?